### PR TITLE
sql: regclass casts use the schema search path

### LINKED ISSUE #13695: sql: REGCLASS casts don't behave properly when the user is root
When the current user is root, virtual tables that expose information about database tables (including `pg_class`) have no restrictions on what tables they show. The query used to look up information in `pg_class` by table name does not qualify the table by the current database, so in the case that two databases have an identically named table, the query will fail due to returning too many results.

### PR DESCRIPTION
**Pull request by jordanlewis**
Previously, casting a string to `regclass` would not respect the schema
search path or current database as a filter to resolve the input table
name. This led to various anomalies, such as the root user (who has
visibility of all tables across all databases) being unable to
regclass-cast a table whose name is repeated in more than one database.

Now, we normalize and resolve the table name the proper way before
querying pg_catalog, and constrain the results of the pg_catalog query
to only those tables in the resolved database.

Also correct a very minor typo in the table descriptor RFC.

Resolves #13695.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/14115)

### CODE REVIEW

File: pkg/sql/testdata/pgoidtype [hunk 1]
```diff
@@ -162,7 +163,31 @@ CREATE DATABASE otherdb
...
+query error table "a" does not exist
 SELECT 'a'::regclass
+
+user root
+
+statement ok
+CREATE TABLE a (id INT PRIMARY KEY, foo STRING)
+
+## root can get the OID of an unqualified table that has a name that's used
+## by a table in a different database, and it'll get the one from the current
```
Comment (knz) [inline pkg/sql/testdata/pgoidtype]: This comment needs more words. I'm confused by "that" and "it".

Comment (jordanlewis) [inline pkg/sql/testdata/pgoidtype]: Yeah this was awfully written. I redid it using an example.

File: pkg/sql/testdata/pgoidtype [hunk 2]
```diff
@@ -162,7 +163,34 @@ CREATE DATABASE otherdb
...
+user root
+
+statement ok
+CREATE TABLE a (id INT PRIMARY KEY, foo STRING)
+
+## There is now a table named 'a' in both the database 'otherdb' and the
+## database 'test'. The following query shows that the root user can still
+## determine the OID of the table 'a' by using a regclass cast, despite the
+## fact that the root user has visibility into both of the tables. The 'a' that
+## gets selected should be the 'a' that exists in the current database.
```
Comment (knz) [inline pkg/sql/testdata/pgoidtype]: But that's not what the code does, right? I think the sentence is "the 'a' that gets selected is the first 'a' that can be found in either the current database or (next) any database in the search path".

Comment (jordanlewis) [inline pkg/sql/testdata/pgoidtype]: I was describing what is going on in this particular test case.

Comment (knz) [inline pkg/sql/testdata/pgoidtype]: Oh I see. Do you have also a test that looks into the search path?
(I think that should happen for regular users too)

**Comment by nvanbenschoten** [comment]
Reviewed 1 of 1 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-Kf9lQD4G7px00z3zgLC:buxq6sp) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
> ```Go
> 				}
> 				info := regTypeInfos[typ]
> 				// Determining the table's OID requires joining against the databases
> ```

Can we just parameterize the WHERE clause in a more specific helper? I hate to see all this duplicated code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115)*

**Comment by jordanlewis** [comment]
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-KfOL5z7gzp4vUXjk0zx:b-4s8okx) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can we just parameterize the WHERE clause in a more specific helper? I hate to see all this duplicated code.
</blockquote></details>

I resisted that at first because it's pretty gross, but maybe it's less gross than before now? I feel a bit wary that this method is duplicating the AST. I think ideally we could redo this to take an AST instead of a string, but I'm not going to do that right now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115)*

**Comment by nvanbenschoten** [comment]
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-KfOlcUJ3YvTV82RWBo_:bfl3szn) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I resisted that at first because it's pretty gross, but maybe it's less gross than before now? I feel a bit wary that this method is duplicating the AST. I think ideally we could redo this to take an AST instead of a string, but I'm not going to do that right now.
</blockquote></details>

I'm not sure I fully understand what you mean by "duplicating the AST".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115)*

**Comment by jordanlewis** [comment]
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-KfOmmAugtF-DcUCvHkR:bqjafk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm not sure I fully understand what you mean by "duplicating the AST".
</blockquote></details>

I didn't explain very clearly, but by adding more ad-hoc and more ways to parameterize a select, ISTM that you lose some readability and you might be better off passing in a hand-constructed expression tree. That was the argument I made to myself for why it might be better to duplicate the code. I'm happy where it is after your suggestion as well, though. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115)*

**Comment by nvanbenschoten** [comment]
Review status: 1 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-KfOr9s2NCnb_7P_ocpN:b-df18ob) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I didn't explain very clearly, but by adding more ad-hoc and more ways to parameterize a select, ISTM that you lose some readability and you might be better off passing in a hand-constructed expression tree. That was the argument I made to myself for why it might be better to duplicate the code. I'm happy where it is after your suggestion as well, though. What do you think?
</blockquote></details>

Ah ok, I think I get what you're saying now and it makes complete sense. I didn't notice the JOIN when I recommended that. Perhaps we should revert to something like what we had before, but introduce a `func queryOid(ctx *EvalContext, sql string, args ...interface{}) (Datum, error)` function. This could handle all of the `MultipleResultsError` and `*DOid` stuff. Then we could have a `queryOidWithRegType` function which is used in most spots and does the `regTypeInfos` mapping.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115)*

**Comment by knz** [comment]
LGTM modulo a possibly missing test that also exercises the search path.

---

Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-:-KfRcyoN3rEvq3W6zehN:bh4qy4g)*

**Comment by jordanlewis** [comment]
Added the test.

---

Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/eval.go, line 2224 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf9lQD4G7px00z3zgLB:-KfS_3IBYLOHHFerKI-R:b-iob1cj) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/parser/eval.go#L2224)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Ah ok, I think I get what you're saying now and it makes complete sense. I didn't notice the JOIN when I recommended that. Perhaps we should revert to something like what we had before, but introduce a `func queryOid(ctx *EvalContext, sql string, args ...interface{}) (Datum, error)` function. This could handle all of the `MultipleResultsError` and `*DOid` stuff. Then we could have a `queryOidWithRegType` function which is used in most spots and does the `regTypeInfos` mapping.
</blockquote></details>

That's a pretty good suggestion. I tried it out though and in practice, it is pretty much the same amount of duplicated code as the first approach I took. All of the error handling needs the `regTypeInfos` value in order to construct a readable message, so the `queryOid` function ends up being pretty much just the call to `QueryRow`. I'm going to leave this alone for now. If we need to continue increasing the complexity here, I think we should switch to constructing an expression tree by hand.

---

*[pkg/sql/testdata/pgoidtype, line 182 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-Kf83o71U1weW279a_NR-r4-182:-KfSaKp2pPf2SW0am-tu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14161143b1ca55cedd6007de496c267b606030a6/pkg/sql/testdata/pgoidtype#L182)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Oh I see. Do you have also a test that looks into the search path?
(I think that should happen for regular users too)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14115#-:-KfS_1hHFgXAV5mWY9x7:b-deoj94)*

**Pull request merged**